### PR TITLE
Enable solution filtering with -graph option

### LINKED ISSUE #7058: Use of NuGet static-graph restore introduces prebuilts in source-build
This is related to the use of the `RestoreUseStaticGraphEvaluation` option which was added in https://github.com/dotnet/msbuild/pull/6905.  This option does not appear to work well with the [`MSBuild.SourceBuild.slnf`](https://github.com/dotnet/msbuild/blob/main/MSBuild.SourceBuild.slnf).  All of the projects in the main solution appear to get restored instead of only the ones in the filter.  Therefore msbuild has numerous prebuilts coming from various projects that are intended to be excluded from source-build.

[source-build msbuild binlog](https://github.com/dotnet/msbuild/files/7582838/2.xx.sourcebuild.zip)

Related to https://github.com/dotnet/source-build/issues/2626

### PR DESCRIPTION
**Pull request by Forgind**
Fixes #7058

### Context
Solution filters operate in MSBuild by modifying the list of projects that "should build." This list is normally constructed in SolutionProjectGenerator, but for -graph builds, it's in GraphBuilder instead. This was hard to notice because it didn't actually cause a build failure because the slnf tries to get out of the way as quickly as it can when you try to build it, which meant the graph build simply built the full sln. This corrects the slnf filtering behavior with -graph.

### Changes Made
Filter the list of projects that "should build" when using -graph.

### Testing
I built MSBuild.dev.slnf before and after the change as well as with and without graph. Eyeballing it, there were about the same number of projects built with the change and -graph as without the change and without -graph. There were substantially more with -graph but without the change.

### CODE REVIEW

File: eng/Build.props [hunk 1]
```diff
@@ -3,7 +3,7 @@
 <Project>
 
   <PropertyGroup>
-    <RestoreUseStaticGraphEvaluation Condition="'$(DotNetBuildFromSource)' != 'true'">true</RestoreUseStaticGraphEvaluation>
+    <RestoreUseStaticGraphEvaluation>true</RestoreUseStaticGraphEvaluation>
```
Comment (jeffkl) [inline eng/Build.props:6]: You’ll need to leave this as-is until your change ships in a build used by Source Build right?

**Review by jeffkl — APPROVED** [review]

File: eng/Build.props (hunk 1 above)
Comment (Forgind) [reply eng/Build.props:6]: If Source Build gets this change in Build.props, wouldn't it also get the fix?

Comment (jeffkl) [reply eng/Build.props:6]: Yeah but source build doesn’t use current MSBuild to build (as far as I know). I think it uses our build.cmd which uses a pretty recent version but will lag behind a little until this change is committed and inserted to VS and .NET SDK. Right?

Comment (Forgind) [reply eng/Build.props]: I know very little about source build, so I have no idea. Pushed the change 🙂 Thanks!

Comment (jeffkl) [reply eng/Build.props]: It looks like it runs `build.sh`:
https://github.com/dotnet/msbuild/blob/764b54b73225a14353e2a1a10e72ae8febb779e0/eng/common/templates/steps/source-build.yml#L71

`build.sh` installs the .NET SDK version based on what's in `global.json`:
https://github.com/dotnet/msbuild/blob/764b54b73225a14353e2a1a10e72ae8febb779e0/eng/Versions.props#L34

So until we're using a version of .NET SDK with this change, we'll need to leave Static Graph restore off for Source Build.  Might be worth opening a tracking issue to do that if you haven't already.

**Comment by jeffkl** [comment]
Thanks for fixing this!

File: eng/Build.props (hunk 1 above)
Comment (rainersigwald) [reply eng/Build.props]: The version of MSBuild used in source-build isn't directly controlled by our repo at all--it's a property of sourcebuild itself. The job that runs in our PR checks is only a simulation and I don't think it's sufficient on this front to be a good canary for when we're ready to remove this.

**Review by rainersigwald — CHANGES_REQUESTED** [review]
Please add a test to cover this case. Also, does this preserve the property of our solution filter handling that projects referenced by projects in the `.slnf` are built?

**Comment by Forgind** [comment]
Seems to. I modified MSBuild.Dev.slnf to be:
```
{
  "solution": {
    "path": "MSBuild.sln",
    "projects": [
      "src\\Build.OM.UnitTests\\Microsoft.Build.Engine.OM.UnitTests.csproj",
      "src\\Build.UnitTests\\Microsoft.Build.Engine.UnitTests.csproj"
    ]
  }
}
```

and it built fine (including things other than M.B.E.OM.UT and M.B.E.UT).

**Review by rainersigwald — COMMENTED** [review]
Please don't move code and edit code in the same commit if at all possible; it makes the diff very hard to parse. I manually reassembled this real diff:

```diff
--- 
+++ 
@@ -1,15 +1,51 @@
         /// <summary>
         /// Test that a solution filter file excludes projects not covered by its list of projects or their dependencies.
         /// </summary>
         [Fact]
         public void SolutionFilterFiltersProjects()
         {
             using (TestEnvironment testEnvironment = TestEnvironment.Create())
             {
-                TransientTestFolder folder = testEnvironment.CreateFolder(createFolder: true);              
+                TransientTestFolder folder = testEnvironment.CreateFolder(createFolder: true);
                 TransientTestFolder classLibFolder = testEnvironment.CreateFolder(Path.Combine(folder.Path, "ClassLibrary"), createFolder: true);
                 TransientTestFolder classLibSubFolder = testEnvironment.CreateFolder(Path.Combine(classLibFolder.Path, "ClassLibrary"), createFolder: true);
                 TransientTestFile classLibrary = testEnvironment.CreateFile(classLibSubFolder, "ClassLibrary.csproj",
                     @"<Project>
                   <Target Name=""ClassLibraryTarget"">
                       <Message Text=""ClassLibraryBuilt""/>
@@ -23,12 +59,13 @@
                     @"<Project DefaultTargets=""SimpleProjectTarget"">
                   <Target Name=""SimpleProjectTarget"">
                       <Message Text=""SimpleProjectBuilt""/>
                   </Target>
                   </Project>
                     ");
+
 
                 // Slashes here (and in the .slnf) are hardcoded as backslashes intentionally to support the common case.
                 TransientTestFile solutionFile = testEnvironment.CreateFile(simpleProjectFolder, "SimpleProject.sln",
                     @"
                     Microsoft Visual Studio Solution File, Format Version 12.00
                     # Visual Studio Version 16
@@ -82,16 +119,22 @@
                 SolutionFile solution = SolutionFile.Parse(filterFile.Path);
                 ILoggingService mockLogger = CreateMockLoggingService();
                 ProjectInstance[] instances = SolutionProjectGenerator.Generate(solution, null, null, _buildEventContext, mockLogger);
                 instances.ShouldHaveSingleItem();
 
                 // Check that dependencies are built, and non-dependencies in the .sln are not.
-                MockLogger logger = new MockLogger(output);
+                MockLogger logger = new(output);
                 instances[0].Build(targets: null, new List<ILogger> { logger }).ShouldBeTrue();
                 logger.AssertLogContains(new string[] { "SimpleProjectBuilt" });
                 logger.AssertLogDoesntContain("ClassLibraryBuilt");
+
+                // Test that the same works for graph builds
+                string log = RunnerUtilities.ExecMSBuild(filterFile.Path + " -graph", out bool success);
+                success.ShouldBeTrue();
+                log.ShouldContain("SimpleProjectBuilt");
+                log.ShouldNotContain("ClassLibraryBuild");
             }
         }
 
         [Theory]
         [InlineData(@"
                 {
@@ -218,6 +261,16 @@
                 (sp.ProjectShouldBuild(Path.Combine("src", Path.GetFileName(commandLineUnitTests.Path)!))
                  || sp.ProjectShouldBuild(Path.Combine("src", Path.GetFileName(msbuild.Path)!))
                  || sp.ProjectShouldBuild(Path.Combine("src", "notAProject.csproj")))
                     .ShouldBeFalse();
             }
         }
+
+        private ILoggingService CreateMockLoggingService()
+        {
+            ILoggingService loggingService = LoggingService.CreateLoggingService(LoggerMode.Synchronous, 0);
+            MockLogger logger = new(output);
+            loggingService.RegisterLogger(logger);
+            return loggingService;
+        }
+    }
+}
```

File: src/Build.UnitTests/Construction/SolutionFilter_Tests.cs [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+                TransientTestFolder simpleProjectSubFolder = testEnvironment.CreateFolder(Path.Combine(simpleProjectFolder.Path, "SimpleProject"), createFolder: true);
+                TransientTestFile simpleProject = testEnvironment.CreateFile(simpleProjectSubFolder, "SimpleProject.csproj",
+                    @"<Project DefaultTargets=""SimpleProjectTarget"">
+                  <Target Name=""SimpleProjectTarget"">
+                      <Message Text=""SimpleProjectBuilt""/>
+                  </Target>
+                  </Project>
+                    ");
+
+
```
Comment (rainersigwald) [inline src/Build.UnitTests/Construction/SolutionFilter_Tests.cs:66]: nit:

```suggestion

```

File: src/Build.UnitTests/Construction/SolutionFilter_Tests.cs [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+                MockLogger logger = new(output);
+                instances[0].Build(targets: null, new List<ILogger> { logger }).ShouldBeTrue();
+                logger.AssertLogContains(new string[] { "SimpleProjectBuilt" });
+                logger.AssertLogDoesntContain("ClassLibraryBuilt");
+
+                // Test that the same works for graph builds
+                string log = RunnerUtilities.ExecMSBuild(filterFile.Path + " -graph", out bool success);
+                success.ShouldBeTrue();
+                log.ShouldContain("SimpleProjectBuilt");
+                log.ShouldNotContain("ClassLibraryBuild");
```
Comment (rainersigwald) [inline src/Build.UnitTests/Construction/SolutionFilter_Tests.cs:134]: I think it would be fairly difficult to understand why this test fails when it fails for only the graph case. Did you find that to be true when you ran this without the fix?

There's a bunch of setup code here so a totally separate test doesn't seem justified, but at the very least a `graphBuild: true/false` theory would help.

File: src/Build.UnitTests/Construction/SolutionFilter_Tests.cs [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+                instances.ShouldHaveSingleItem();
+
+                // Check that dependencies are built, and non-dependencies in the .sln are not.
+                MockLogger logger = new(output);
+                instances[0].Build(targets: null, new List<ILogger> { logger }).ShouldBeTrue();
+                logger.AssertLogContains(new string[] { "SimpleProjectBuilt" });
+                logger.AssertLogDoesntContain("ClassLibraryBuilt");
+
+                // Test that the same works for graph builds
+                string log = RunnerUtilities.ExecMSBuild(filterFile.Path + " -graph", out bool success);
```
Comment (rainersigwald) [inline src/Build.UnitTests/Construction/SolutionFilter_Tests.cs:131]: Why does this invoke `msbuild.exe` rather than using the API like the one above does?

File: src/Build.UnitTests/Construction/SolutionFilter_Tests.cs (hunk 3 above)
Comment (Forgind) [reply src/Build.UnitTests/Construction/SolutionFilter_Tests.cs:132]: I was waffling on this a little. We have some tests in the theory model and some more like this where, after setting up, it runs multiple tests in a row on the same thing. I thought this was more appropriate for the latter because most of it is boring setup. It's hard for me to say how debuggable it is when I know the answer ahead of time. My first priority is getting something that will fail if someone breaks this. I honestly doubt anyone other than me will be trying to debug this in the near future...

File: src/Build.UnitTests/Construction/SolutionFilter_Tests.cs (hunk 4 above)
Comment (Forgind) [reply src/Build.UnitTests/Construction/SolutionFilter_Tests.cs:129]: Using the API is better, but this was something I mostly knew how to do already. If you feel strongly, I can try to figure out the graph API. I would guesstimate an hour or two.

File: src/Build.UnitTests/Construction/SolutionFilter_Tests.cs [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+                ");
+                Directory.GetCurrentDirectory().ShouldNotBe(Path.GetDirectoryName(filterFile.Path));
+                if (graphBuild)
+                {
+                    ProjectCollection projectCollection = testEnvironment.CreateProjectCollection().Collection;
+                    MockLogger logger = new();
+                    logger.Verbosity = LoggerVerbosity.Diagnostic;
+                    projectCollection.RegisterLogger(logger);
+                    ProjectGraphEntryPoint entryPoint = new(filterFile.Path, new Dictionary<string, string>());
+                    ProjectGraph graphFromSolution = new(entryPoint, projectCollection);
```
Comment (rainersigwald) [inline src/Build.UnitTests/Construction/SolutionFilter_Tests.cs:127]: Please at least comment that it's sufficient to construct the graph and it doesn't require building it.

**Review by rainersigwald — APPROVED** [review]
